param.go

```diff
@@ -129,16 +137,111 @@ func newParamList(ctype reflect.Type) (paramList, error) {
 		}
 		pl.Params = append(pl.Params, p)
 	}
+
 	return pl, nil
 }
 
+func (pl paramList) Build(*Container) (reflect.Value, error) {
+	panic("paramList.Build() must never be called")
+}
+
+// BuildParams returns an ordered list of values which may be passed directly
+// to the underlying constructor.
+func (pl paramList) BuildParams(c *Container) ([]reflect.Value, error) {
+	args := make([]reflect.Value, len(pl.Params))
+	for i, p := range pl.Params {
+		var err error
+		args[i], err = p.Build(c)
+		if err != nil {
+			return nil, errWrapf(err, "could not build argument %d for constructor %v", i, pl.ctype)
+		}
+	}
+	return args, nil
+}
+
 // newParam builds a param from the given type. If the provided type is a
 // dig.In struct, an paramObject will be returned.
 func newParam(t reflect.Type) (param, error) {
-	if IsIn(t) {
+	switch {
+	case IsIn(t):
 		return newParamObject(t)
+	case embedsType(t, _inPtrType):
+		return nil, fmt.Errorf(
+			"%v embeds *dig.In which is not supported, embed dig.In value instead", t)
+	case t.Kind() == reflect.Ptr && IsIn(t.Elem()):
+		return nil, fmt.Errorf(
+			"dependency %v is a pointer to dig.In, use value type instead", t)
+	case IsOut(t) || (t.Kind() == reflect.Ptr && IsOut(t.Elem())) || embedsType(t, _outPtrType):
+		return nil, fmt.Errorf("cannot depend on result objects: %v embeds a dig.Out", t)
+	default:
+		return paramSingle{Type: t}, nil
 	}
-	return paramSingle{Type: t}, nil
+}
+
+func (sp paramSingle) Build(c *Container) (reflect.Value, error) {
+	k := key{name: sp.Name, t: sp.Type}
+	if v, ok := c.cache[k]; ok {
+		return v, nil
+	}
+
+	n, ok := c.nodes[k]
+	if !ok {
+		// Unlike in the fallback case below, if a user makes an error requesting
+		// a mixed type for an optional parameter, a good error message "did you mean X?"
+		// will not be used and dig will return zero value.
```

Comment:
What's a "mixed type" in this context? The only references I can Google are PHP-specific and don't make much sense for Go.

Do we even need a comment explaining that optional params are not required?

Reply:
Not entirely sure. This was copied verbatim. Reworded to make more sense.

Reply:
requesting *foo when foo is available
requesting foo when *foo is available is what i labeled "mixed type".

If you can figure out a better verbage - go for it. I couldn't